### PR TITLE
executor: notify fetch goroutines to exit when IndexLookUpExecutor close

### PR DESCRIPTION
**Pull request by tiancaiamao**
For a query `select * from t where id > xxx limit 1000` on a table whose size is 500 million,
the query return but not really finished, the next query would block for 1 minute.

After some investigation, I find `Close()` function hang, and here is the goroutine stack:

```
3 @ 0x42cafa 0x42cbde 0x404e31 0x404aa5 0xd83e63 0xd660a4 0xd660a4 0xd660a4 0xd443f8 0xf02583 0xefdbfa 0xefedd2 0xefadab 0xef980c 0xf09f78 0x459ae1
#	0xd83e62	github.com/pingcap/tidb/executor.(*IndexLookUpExecutor).Close+0x62	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/new_distsql.go:498
#	0xd660a3	github.com/pingcap/tidb/executor.(*baseExecutor).Close+0x53		/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/executor.go:123
#	0xd660a3	github.com/pingcap/tidb/executor.(*baseExecutor).Close+0x53		/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/executor.go:123
#	0xd660a3	github.com/pingcap/tidb/executor.(*baseExecutor).Close+0x53		/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/executor.go:123
#	0xd443f7	github.com/pingcap/tidb/executor.(*recordSet).Close+0x37		/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/adapter.go:82
#	0xf02582	github.com/pingcap/tidb/server.(*tidbResultSet).Close+0x32		/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/driver_tidb.go:299
#	0xefdbf9	github.com/pingcap/tidb/server.(*clientConn).writeResultset+0xaf9	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:805
#	0xefedd1	github.com/pingcap/tidb/server.(*clientConn).handleStmtExecute+0x431	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn_stmt.go:174
#	0xefadaa	github.com/pingcap/tidb/server.(*clientConn).dispatch+0x80a		/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:483
#	0xef980b	github.com/pingcap/tidb/server.(*clientConn).Run+0x13b			/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:361
#	0xf09f77	github.com/pingcap/tidb/server.(*Server).onConn+0x147			/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/server.go:236
```

@shenli @hanfei1991

### CODE REVIEW

File: executor/new_distsql.go [hunk 1]
```diff
@@ -307,7 +307,6 @@ type IndexLookUpExecutor struct {
 
 // Open implements the Executor Open interface.
 func (e *IndexLookUpExecutor) Open() error {
```
Comment (shenli) [inline executor/new_distsql.go]: This is not called? @hanfei1991

File: executor/builder.go [hunk 2]
```diff
@@ -1137,6 +1137,7 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
 		columns:      is.Columns,
 		handleCol:    handleCol,
 		priority:     b.priority,
+		finish:       make(chan struct{}),
```
Comment (ngaut) [inline executor/builder.go]: s/finish/finished/

**Comment by ngaut** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @shenli @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**